README.md
# Firefox PKCS11 Loader

![European Regional Development Fund](https://github.com/e-gov/RIHA-Frontend/raw/master/logo/EU/EU.png "European Regional Development Fund - DO NOT REMOVE THIS IMAGE BEFORE 05.03.2020")

 * License: LGPL 2.1
 * &copy; Estonian Information System Authority

## Building
[![Build Status](https://travis-ci.org/open-eid/firefox-pkcs11-loader.svg?branch=master)](https://travis-ci.org/open-eid/firefox-pkcs11-loader)
[![Build Status](https://ci.appveyor.com/api/projects/status/github/open-eid/firefox-pkcs11-loader?branch=master&svg=true)](https://ci.appveyor.com/project/open-eid/firefox-pkcs11-loader)
        
### Windows/OSX/Ubuntu

1. Fetch the source

        git clone https://github.com/open-eid/firefox-pkcs11-loader
        cd firefox-pkcs11-loader

2. Configure

        mkdir build
        cd build
        cmake ..

3. Build

        make extension
        make installer

4. Install the package and execute Firefox

## Support
Official builds are provided through official distribution point [installer.id.ee](https://installer.id.ee). If you want support, you need to be using official builds. Contact our support via [www.id.ee](http://www.id.ee) for assistance.

Source code is provided on "as is" terms with no warranty (see license for more information). Do not file Github issues with generic support requests.


RELEASE-NOTES.md
Firefox PKCS11 Loader library [3.13.6](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.6) release notes
--------------------------------------
- Add jq debian dependency (#28)

Firefox PKCS11 Loader library [3.13.5](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.5) release notes
--------------------------------------
- Create policies.json on ubuntu (#26)
- Change WIX CustomAction execution conditions (#25)

Firefox PKCS11 Loader library [3.13.4](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.4) release notes
--------------------------------------
- Distribute Mozilla AMO signed extensions

Firefox PKCS11 Loader library [3.13.3](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.3) release notes
--------------------------------------
- Unload old IDEMIA PKCS11

Firefox PKCS11 Loader library [3.13.2](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.2) release notes
--------------------------------------
- Load IDEMIA PKCS11

Firefox PKCS11 Loader library [3.13.1](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.1) release notes
--------------------------------------
- Fix Firefox 64 Windows 64 bit support

Firefox PKCS11 Loader library [3.13.0](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.13.0) release notes
--------------------------------------
- Firefox 58 support

[Full Changelog](https://github.com/open-eid/firefox-pkcs11-loader/compare/v3.12.1...v3.13.0)

Firefox PKCS11 Loader library [3.12.1](https://github.com/open-eid/firefox-pkcs11-loader/releases/tag/v3.12.1) release notes
--------------------------------------
- Use OpenSC on macOS
- Build improvements

[Full Changelog](https://github.com/open-eid/firefox-pkcs11-loader/compare/v3.12.0...v3.12.1)


Firefox PKCS11 Loader version 3.12.0 release notes
--------------------------------------
- Fix compatibility with Firefox 47
- The Firefox authentication module is now signed by Mozilla when publishing it to Add-Ons store.  


Firefox PKCS11 Loader version 3.11.1 release notes
--------------------------------------
- Add cmake target to create windows/mac installer


Esteid-pkcs11-module-loader version 3.11.0 release notes
--------------------------------------
- Cleanup JavaScript namespace pollution


Esteid-pkcs11-module-loader version 3.8.1 release notes
--------------------------------------
- Added one-pin module to improve authentication in Firefox web browser in OSX environment.


webextension/background.js
/*
 * Firefox PKCS11 loader
 *
 * This is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

(async function () {
  if(typeof browser.pkcs11 === 'undefined')
    return;
  async function load(modname) {
    try {
      var isInstalled = await browser.pkcs11.isModuleInstalled(modname);
      if(isInstalled) {
        console.log("module installed: " + modname);
        return;
      }
      const PKCS11_PUB_READABLE_CERT_FLAG = 0x1<<28;
      await browser.pkcs11.installModule(modname, PKCS11_PUB_READABLE_CERT_FLAG);
      console.log("Loaded module " + modname);
    } catch(e) {
      console.error("Unable to load module: ", e);
    }
  }
  async function unload(modname) {
    try {
      var isInstalled = await browser.pkcs11.isModuleInstalled(modname);
      if(!isInstalled) {
        console.log("module is not installed: " + modname);
        return;
      }
      await browser.pkcs11.uninstallModule(modname);
      console.log("Unloaded module " + modname);
    } catch(e) {
      console.error("Unable to unload module: ", e);
    }
  }
  load("onepinopenscpkcs11");
  unload("idemiaawppkcs11");
})();


chrome/content/pkcs11-loader.js
/*
 * Estonian ID card plugin for web browsers
 *
 * This is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */
 
(function () {
  const EstEidModName = "Estonian ID Card";
  var Ci = Components.interfaces;
  var Cc = Components.classes;
  var pkcs11db = Cc["@mozilla.org/security/pkcs11moduledb;1"].createInstance(Ci.nsIPKCS11ModuleDB);
  var pkcs11 = Cc["@mozilla.org/security/pkcs11;1"].getService(Ci.nsIPKCS11);
  var httpHandler = Cc["@mozilla.org/network/protocol;1?name=http"].getService(Ci.nsIHttpProtocolHandler);
  var log = function(message) { console.log(message); }

  log("Loading extension " + EstEidModName);

  var moduleDll = "";
  switch (httpHandler.platform) {
  case "Windows":
    moduleDll = "onepin-opensc-pkcs11.dll";
    break
  case "Macintosh":
    moduleDll = "/Library/OpenSC/lib/onepin-opensc-pkcs11.so";
    break;
  default:
    [ "/usr/lib64/onepin-opensc-pkcs11.so",
      "/usr/lib64/opensc-pkcs11.so",
      "/usr/lib/x86_64-linux-gnu/onepin-opensc-pkcs11.so",
      "/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so",
      "/usr/lib/i386-linux-gnu/onepin-opensc-pkcs11.so",
      "/usr/lib/i386-linux-gnu/opensc-pkcs11.so",
      "/usr/lib/onepin-opensc-pkcs11.so",
      "/usr/lib/opensc-pkcs11.so"
    ].some(function(path) {
      log(path);
      var f = Cc["@mozilla.org/file/local;1"].createInstance(Ci.nsIFile);
      f.initWithPath(path);
      if (f.exists())
        moduleDll = path;
      return f.exists();
    });
  }

  try {
    var module = pkcs11db.findModuleByName(EstEidModName);
    if (module.libName == moduleDll) {
      log("Module already initialized " + module.name + " (" + module.libName + ")");
      return;
    }
    try {
      log("Removing module " + module.name + " (" + module.libName + ")");
      pkcs11.deleteModule(module.name);
    }
    catch(e) {
      log("Unable to remove module:" + e);
    }
  }
  catch(e) {
  }

  try {
    log("Loading module " + EstEidModName + " (" + moduleDll + ")");
    //Stored certs can be read off the token w/o logging in
    const PKCS11_PUB_READABLE_CERT_FLAG = 0x1<<28;
    pkcs11.addModule(EstEidModName, moduleDll, PKCS11_PUB_READABLE_CERT_FLAG, 0);
  }
  catch(e) {
    log("Unable to load module: " + e);
  }
})();
